Print values (but not strings) when STRIP_ERROR_MESSAGES is defined

changjishi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40700 Print values (but not strings) when STRIP_ERROR_MESSAGES is defined**

Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: 

Use timer_test.cc as a test file: compared the binary size and error message output

Version in diff: 441752 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

unknown file: Failure
C++ exception with description "
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fadfbbd9be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)


time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "1594157690, string, 2.12345, 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fadfbbd9be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Current error message with strip: 442040 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

unknown file: Failure
C++ exception with description ", 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fd058df5be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)

time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "1594173782, string, 2.12345, , 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fd058df5be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Current error message without strip: 442104 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

nknown file: Failure
C++ exception with description "false INTERNAL ASSERT FAILED at "../caffe2/core/timer_test.cc":43, please report a bug to PyTorch. 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f4086e53be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)

time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "false INTERNAL ASSERT FAILED at "../caffe2/core/timer_test.cc":77, please report a bug to PyTorch. 1594174314hello worldnot important informationstring2.12345
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f4086e53be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

Commit:
Print values (but not strings) when STRIP_ERROR_MESSAGES is defined

Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

ezyang commented:
In private correspondence, we discussed that probably we can simplify the template code, rather than just copy pasting the implementation of c10::str

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

ezyang commented on c10/util/StringUtil.h:
```diff
@@ -87,6 +134,14 @@ inline decltype(auto) str(const Args&... args) {
+// if an argument is const char*, it will not show up in the returned message.
+// This is to optimize build size.
+template <typename... Args>
+inline decltype(auto) error_value(const Args&... args) {
```
You probably want to make this unconditionally accept some number of leading arguments; otherwise you'll lose basic diagnostics like " INTERNAL ASSERT FAILED at"

changjishi replied:
Had an offline discussion, because file, line information is already contained in error, we decided to only pass error arguments to message.

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

ezyang reviewed (approved):
This looks great. Let's do a check for the internal Android build size and it's good to go.

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: 

Use timer_test.cc as a test file: compared the binary size and error message output

Version in diff: 441752 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

unknown file: Failure
C++ exception with description "
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fadfbbd9be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)


time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "1594157690, string, 2.12345, 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fadfbbd9be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Current error message with strip: 442040 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

unknown file: Failure
C++ exception with description ", 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fd058df5be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)

time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "1594173782, string, 2.12345, , 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fd058df5be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Current error message without strip: 442104 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

nknown file: Failure
C++ exception with description "false INTERNAL ASSERT FAILED at "../caffe2/core/timer_test.cc":43, please report a bug to PyTorch. 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f4086e53be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)

time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "false INTERNAL ASSERT FAILED at "../caffe2/core/timer_test.cc":77, please report a bug to PyTorch. 1594174314hello worldnot important informationstring2.12345
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f4086e53be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]

Commit:
Update on "Print values (but not strings) when STRIP_ERROR_MESSAGES is defined"


Summary: Currently, to save binary size on mobile builds, we completely remove all error message strings. This can make it difficult to debug problems as some parts of the error output (namely various runtime values which would have been incorporated into the error message string) get suppressed entirely. It is not so convenient to attach gdb to a mobile program. So add a variant of c10::str, which will not print any string or char* into the error message

Test Plan: 

Use timer_test.cc as a test file: compared the binary size and error message output

Version in diff: 441752 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

unknown file: Failure
C++ exception with description "
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fadfbbd9be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)


time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "1594157690, string, 2.12345, 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fadfbbd9be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Current error message with strip: 442040 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

unknown file: Failure
C++ exception with description ", 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fd058df5be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)

time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "1594173782, string, 2.12345, , 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fd058df5be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Current error message without strip: 442104 (binary size)

TORCH_INTERNAL_ASSERT(false) 

output:

nknown file: Failure
C++ exception with description "false INTERNAL ASSERT FAILED at "../caffe2/core/timer_test.cc":43, please report a bug to PyTorch. 
Exception raised from TestBody at ../caffe2/core/timer_test.cc:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f4086e53be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)

time_t current_time = time(NULL);
const char* ca = "hello world";
std::string str = "string";
double d = 2.12345;
TORCH_INTERNAL_ASSERT(false, current_time, ca, "not important information", str, d)

output: 

unknown file: Failure
C++ exception with description "false INTERNAL ASSERT FAILED at "../caffe2/core/timer_test.cc":77, please report a bug to PyTorch. 1594174314hello worldnot important informationstring2.12345
Exception raised from TestBody at ../caffe2/core/timer_test.cc:77 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f4086e53be0 in /data/users/cjshi/pytorch/build/lib/libc10.so)



Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22284218](https://our.internmc.facebook.com/intern/diff/D22284218)

[ghstack-poisoned]
